web/src/ui/timeline/TimelineView.tsx

```diff
@@ -95,26 +133,31 @@ const TimelineView = () => {
 			)
 		}
 	}, [focused, client, roomCtx, room, timeline])
+
 	useEffect(() => {
-		const topElem = topRef.current
-		if (!topElem || !room.hasMoreHistory) {
+		if (!room.hasMoreHistory || room.paginating) {
+			return
+		}
+
+		const firstItem = virtualizer.getVirtualItems()[0]
+
+		// Load history if there is none
+		if (!firstItem) {
+			loadHistory()
 			return
 		}
-		const observer = new IntersectionObserver(entries => {
-			if (entries[0]?.isIntersecting && room.paginationRequestedForRow !== prevOldestTimelineRow.current) {
-				room.paginationRequestedForRow = prevOldestTimelineRow.current
-				loadHistory()
-			}
-		}, {
-			root: topElem.parentElement!.parentElement,
-			rootMargin: "0px",
-			threshold: 1.0,
-		})
-		observer.observe(topElem)
-		return () => observer.unobserve(topElem)
-	}, [room, room.hasMoreHistory, loadHistory, oldestTimelineRow])
-
-	let prevEvt: MemDBEvent | null = null
+
+		// Load more history when the virtualiser loads the last item
+		if (firstItem.index == 0) {
+			console.log("Loading more history...")
+			loadHistory()
+			return
+		}
+	}, [
+		room.hasMoreHistory, loadHistory,
+		virtualizer.getVirtualItems(),
```

Comment:
This effect is now ran on every render because `loadHistory` changes every time (`useCallback`s are not there just for fun). Eslint is also complaining about missing dependencies

Reply:
This seems to be run on every render *anyway* due to `virtualizer.getVirtualItems()`. Not sure what to do about this, as it's the suggested way to achieve this effect in https://tanstack.com/virtual/latest/docs/framework/react/examples/infinite-scroll, and from what I can see, the actual result of `firstItem` doesn't change.